src/Compilers/CSharp/Portable/Symbols/Metadata/PE/ExternalAnnotations.cs

```diff
@@ -0,0 +1,111 @@
+        {
+            var metadata = moduleSymbol.Module.MetadataReader;
+            var builder = PooledStringBuilder.GetInstance();
+
+            if (handle.Kind == HandleKind.MethodDefinition)
+            {
+                Add((MethodDefinitionHandle)handle, metadata, builder, moduleSymbol);
+            }
+
+            return builder.ToStringAndFree();
+        }
+
+        private static readonly ImmutableArray<bool> skip = default;
+
+        static ImmutableArray<ImmutableArray<bool>> Parameters(params ImmutableArray<bool>[] values)
+        {
+            return values.ToImmutableArray();
+        }
+
+        static ImmutableArray<bool> Nullable(params bool[] values)
+        {
+            return values.ToImmutableArray();
+        }
+
+        internal static ImmutableArray<bool> GetExtraAnnotations(string key, int ordinal)
+        {
+            if (!Annotations.TryGetValue(key, out var flags))
+            {
+                return default;
+            }
+
+            return flags[ordinal];
+        }
+
+        private static void Add(MethodDefinitionHandle methodHandle, MetadataReader metadata, PooledStringBuilder builder, PEModuleSymbol moduleSymbol)
+        {
+            var metadataDecoder = new MetadataDecoder(moduleSymbol);
+            SignatureHeader signatureHeader;
+            BadImageFormatException mrEx;
+            ParamInfo<TypeSymbol>[] methodParams = metadataDecoder.GetSignatureForMethod(methodHandle, out signatureHeader, out mrEx, setParamHandles: false);
+
+            MethodDefinition method = metadata.GetMethodDefinition(methodHandle);
+
+            TypeSymbol type1 = methodParams[0].Type;
+            Add(type1, builder);
+
+            builder.Builder.Append(' ');
+
+            TypeDefinition type = metadata.GetTypeDefinition(method.GetDeclaringType());
+            Add(type, metadata, builder);
+            builder.Builder.Append('.');
+
+            string name = metadata.GetString(method.Name);
+            builder.Builder.Append(name);
+            builder.Builder.Append('(');
+
+            for (int i = 1; i < methodParams.Length; i++)
+            {
+                Add(methodParams[i].Type, builder);
+                if (i < methodParams.Length - 1)
+                {
+                    builder.Builder.Append(", ");
+                }
+            }
+
+            builder.Builder.Append(')');
+
+            // PROTOTYPE(NullableReferenceTypes): Many cases are not yet handled
+            // generic type args
+            // ref kind
+            // 'this'
+        }
+
+        private static void Add(TypeSymbol type1, PooledStringBuilder builder) =>
+                        builder.Builder.Append(type1.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat.RemoveMiscellaneousOptions(SymbolDisplayMiscellaneousOptions.UseSpecialTypes)));
+
+        private static void Add(TypeDefinition type, MetadataReader metadata, PooledStringBuilder builder)
```

Comment:
Consider changing parameter to `StringBuilder`.

Reply:
The methods depend on `StringBuilder` only, not on `PooledStringBuilder`.

Reply:
`PooledStringBuilder` is better for resource management.

Reply:
@cston was suggesting that the `Add` methods only get passed a `StringBuilder`. The caller would still use `PooledStringBuilder`.
This simplifies the code inside the `Add` methods. So we get all the benefits.